bottombar.py

```diff
@@ -20,152 +20,303 @@
-        b'\033[1;%dr' # set scroll region
-        b'\0338' # restore cursor and attributes
-        % (size.lines-1))
-      self.size = size
-      force = True
-    if force:
-      self.write(
-        b'\0337' # save cursor and attributes
-        b'\033[%d;1H' # move cursor to bottom row, first column
-        b'\033[2K' # clear entire line
-        b'\033[?7l' # disable line wrap
-        b'\033[0m' # clear attributes
-        b'%s' # print bar
-        b'\033[?7h' # enable line wrap
-        b'\0338' # restore cursor and attributes
-        % (size.lines, self.format(*self.args, width=size.columns).encode(self.encoding)))
-
-  def __exit__(self, *exc):
-    if self.size is None:
-      raise RuntimeError('BottomBar has not yet been entered')
-    while self.handles:
-      self.handles.pop().close()
-    self.write(
-      b'\0337' # save cursor position
-      b'\033[%d;1H' # move cursor to bottom row, first column
-      b'\033[K' # clear entire line
-      b'\033[r' # reset scroll region
-      b'\0338' # restore cursor position
-      % self.size.lines)
-    self.size = None
-
-
-# vim:sw=2:sts=2:et
+    def __init__(self):
+        self.__nleft = 0
+        self.__items = []
+
+    def __bool__(self):
+        return bool(self.__items)
+
+    def add(self, text, *, right=False, label=None):
+        '''Create a bar item and add it to the bar.
+
+        The new item is added to the left or right stack depending on the
+        `right` argument. The new BarItem instance is returned for in place
+        modification and removal.'''
+
+        c = BarItem(text, label)
+        self.__items.insert(self.__nleft, c)
+        if not right:
+            self.__nleft += 1
+        return c
+
+    def remove(self, item):
+        '''Remove a bar item.'''
+
+        index = self.__items.index(item)
+        del self.__items[index]
+        if index < self.__nleft:
+            self.__nleft -= 1
+
+    def format(self, length):
+        '''Generate bar string of given length.
+
+        Depending on the total size of the bar items and the target bar length,
+        the formatted string will show labeled items as | label: text | if
+        there is sufficient space, or simply | text | if there is there is
+        sufficient space for all text, or | t.. | if (some) items require
+        shortening. The returned string always matches the requested length.'''
+
+        itemsep = ' | '
+        labelsep = ': '
+        nitems = len(self.__items)
+        maxchars = length - len(itemsep) * (nitems - 1)
+        texts = [str(item.text) for item in self.__items]
+        whitespace = maxchars - sum(map(len, texts))
+        if whitespace < 0:
+            # not all items fit; shorten the longest ones
+            for i in sorted(range(nitems), key=lambda i: len(texts[i])): # argsort
+                maxlen = maxchars // nitems
+                if len(texts[i]) >= maxlen:
+                    texts[i] = texts[i][:maxlen-2] + '..'
```

Comment:
I branched off the situation `maxchars <= 2 * nitems` (rather than negative). This way the first `maxlen >= 2`, and all subsequent ones should be larger or equal.